Sources/JSONUtilities/Errors.swift

```diff
@@ -13,40 +13,54 @@ import Foundation
 
  - Mandatory: The key that failed the decoding
  */
-public enum DecodingError: Error, CustomStringConvertible {
-  case keyNotFound(dictionary: [AnyHashable: Any], key: StringProtocol)
-  case incorrectRawRepresentableRawValue(rawRepresentable: Any.Type, rawValue: JSONRawType)
-  case incorrectTypeInArray(array: JSONArray, expectedType: Any.Type, value: Any)
-  case incorrectTypeInDictionary(dictionary: [AnyHashable: Any], key: StringProtocol, expectedType: Any.Type, value: Any)
-  case conversionFailure(type: Any.Type, value: Any)
+public struct DecodingError: Error, CustomStringConvertible, CustomDebugStringConvertible {
 
-  public var description: String {
-    switch self {
-    case .keyNotFound(_, let key):
-      return "Key not found: \(key)"
-    case .incorrectRawRepresentableRawValue(let rawRepresentable, let rawValue):
-      return "Incorrect rawValue for \(rawRepresentable): \(rawValue)"
-    case .incorrectTypeInArray(_, let expectedType, let value):
-      return "Incorrect type in [\(expectedType)]: \(value)"
-    case .incorrectTypeInDictionary(_, let key, let expectedType, let value):
-      return "Incorrect type in [String: \(expectedType)] for key \"\(key)\": \(value)"
-    case .conversionFailure(let type, let value):
-      return "\(type) failed to convert: \(value)"
-    }
+  public var reason: Reason
+  public let dictionary: [AnyHashable: Any]
+  public let array: JSONArray?
+  public let keyPath: String
+  public let expectedType: Any
+  public let value: Any
+
+  init(dictionary: [AnyHashable: Any], keyPath: StringProtocol, expectedType: Any.Type, value: Any, array: JSONArray? = nil, reason: Reason) {
+    self.dictionary = dictionary
+    self.keyPath = keyPath.string
+    self.expectedType = expectedType
+    self.value = value
+    self.reason = reason
+    self.array = array
   }
 
-  public var name: String {
-    switch self {
+  var reasonDescription: String {
+    switch reason {
     case .keyNotFound:
-      return "Key not found"
+      return "Nothing found"
     case .incorrectRawRepresentableRawValue:
-      return "Incorrect RawRepresentable RawValue"
-    case .incorrectTypeInArray:
-      return "Incorrect type in array"
-    case .incorrectTypeInDictionary:
-      return "Incorrect type in dictionary"
+      return "Incorrect rawValue of \(value) for \(expectedType)"
+    case .incorrectType:
+      return "Incorrect type, expected \(expectedType) found \(value)"
     case .conversionFailure:
-      return "Conversion failure"
+      return "\(expectedType) failed to convert \(value)"
+    }
+  }
+
+  public var description: String {
+    return "Decoding failed at \"\(keyPath)\": \(reasonDescription)"
+  }
+
+  public var debugDescription: String {
+    return "\(description):\n\(array?.description ?? dictionary.description)"
+  }
+
+  public enum Reason: String, CustomStringConvertible {
```

Comment:
Could we have docs for the `Reason`, not 100%  what conversion failure means for example, from the code I guess its a conversion to a primitive but not so clear?